gax/src/main/java/com/google/api/gax/batching/FlowControlEventStats.java

```diff
@@ -0,0 +1,142 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.batching;
+
+import static com.google.api.gax.batching.FlowController.LimitExceededBehavior;
+
+import com.google.api.core.InternalApi;
+import com.google.api.gax.batching.FlowController.FlowControlException;
+import com.google.auto.value.AutoValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Record the statistics of flow control events. Currently this class only captures the last flow
+ * control event. But it could be expanded to record more information in the future.
+ *
+ * <p>If {@link FlowController.LimitExceededBehavior} is {@link LimitExceededBehavior#Block} and
+ * {@link FlowController#reserve(long, long)} takes longer than expected, record the flow control
+ * event with the throttled time. For example:
+ *
+ * <pre>{@code
+ * Stopwatch stopwatch = Stopwatch.createStarted();
+ * flowController.reserve(10, 10);
+ * long elapsed = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+ * // If reserve takes longer than 1 millisecond, mark it as a flow control event
+ * if (elapsed > 1) {
+ *    flowControlEventStats.recordFlowControlEvent(FlowControlEvent.createReserveDelayed(elapsed));
+ * }
+ * }</pre>
+ *
+ * If {@link FlowController.LimitExceededBehavior} is {@link LimitExceededBehavior#ThrowException}
+ * and {@link FlowController#reserve(long, long)} throws a {@link FlowControlException}, record the
+ * flow control event with the exception. For example:
+ *
+ * <pre>{@code
+ * try {
+ *   flowController.reserve(10, 10);
+ * } catch (FlowControlException e) {
+ *   flowControlEventStats.recordFlowControlEvent(FlowControlEvent.createReserveDenied(exception));
+ * }
+ * }</pre>
+ */
```

Comment:
Sorry I shouldve been a bit more clear I was asking for docs from the client dev perspective

Something along the lines of:
This class is populated by FlowController, which will record throttling events. This events can be used to dynamically adjust concurrency. For example:
```
while(true) {
   e = stats.getLastFlowControlEvent();
   if (e != null && e.getTimestampMs() > System.currentMillis() - 1 minute && e.getThrottledTimeInMs() > 1 min) {
     decreaseConcurrency();
   }
   Thread.sleep(10mins);
}```

Reply:
Ah I see 😅 I thought you meant document how to use this class. Will update.